GLI loader WIP

Crisspl opened the pull request.

AnastaZIuk committed:
Small fixes, it still need to consider cube arrays issue

AnastaZIuk committed:
Fix cubemaps' validation, add swizzle checking on writing

AnastaZIuk committed:
Merge branch 'shaders' of https://github.com/Crisspl/IrrlichtBAW into GLIdep

Sync with Criss

AnastaZIuk committed:
Remove some references according to Criss' advice

AnastaZIuk committed:
Merge branch 'shaders' of https://github.com/Crisspl/IrrlichtBAW into GLIdep

Merging fixes

AnastaZIuk committed:
Fix array layers equation, remove some formats

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.cpp:
```diff
@@ -196,6 +223,27 @@ namespace irr
+			case ICPUImageView::ET_CUBE_MAP_ARRAY: return true;
+			default: return false;
+			}
+		}
```
move the definitions to header so they are actually `inline`

devshgraphicsprogramming commented on src/irr/asset/CGLIWriter.cpp:
```diff
@@ -146,7 +155,7 @@ namespace irr
 				for (uint16_t layer = 0; layer < imageInfo.arrayLayers; ++layer)
 				{
 					const uint16_t gliLayer = layersFlag ? layer / 6 : 0;
```
layer still wrong

devshgraphicsprogramming replied:
still wrong

devshgraphicsprogramming replied:
after https://github.com/buildaworldnet/IrrlichtBAW/pull/411/files/3e1da52a9ec2f98eb970baffdf29ea139282b499..426d151665e4292916c1c0a2fd69deaa3416fd05

now is correct.

AnastaZIuk committed:
Adjust GLI dependency to IReadFile and IWriteFile

AnastaZIuk committed:
Add validation signature file, add some formats, fix the equation

AnastaZIuk committed:
Add some copression formats, move functions and define them in GLI writer/loader's header

AnastaZIuk committed:
Add the utility function, fix equation in the writer and a few docs words

The utility function allows you to convert EF_R8_SRGB to EF_R8G8B8_SRGB
and fills values of G and B channels with R value (RRR)

devshgraphicsprogramming commented on src/irr/asset/CGLIWriter.cpp:
```diff
@@ -153,12 +153,28 @@ namespace irr
+				const auto textureGliStride = texture.extent(region->imageSubresource.mipLevel).x;
+				const auto textureGliStrideInBytes = textureGliStride * sizeof(singleChannelByteSize);
+				const auto imgBufferWidth = region->bufferRowLength > 0 ? region->bufferRowLength : region->imageExtent.width;
+				const auto imgBufferWidthInBytes = imgBufferWidth * sizeof(singleChannelByteSize);
```
ekhm, what are you expecting to get as an answer from `sizeof(singleChannelByteSize)` ?

devshgraphicsprogramming commented on src/irr/asset/CGLIWriter.cpp:
```diff
@@ -153,12 +153,28 @@ namespace irr
+					const auto layerData = texture.data(gliLayer, gliFace, region->imageSubresource.mipLevel);
+					const auto sourceData = ptrBeginningOfRegion + (layer * layerSize);
+
+					for (uint32_t yPos = 0; yPos < textureGliImgHeight; ++yPos)
```
you;re missing a for-loop along Z coordinate

devshgraphicsprogramming replied:
Also dont increment `++`, increment by `+=` block dimensions (there's a function that will give you that dimension)

devshgraphicsprogramming replied:
@AnastaZIuk status?

AnastaZIuk replied:
done

devshgraphicsprogramming commented on src/irr/asset/CGLIWriter.cpp:
```diff
@@ -153,12 +153,28 @@ namespace irr
+						(
+							reinterpret_cast<uint8_t*>(layerData) + (yPos * textureGliStrideInBytes),
+							sourceData + (yPos * imgBufferWidthInBytes),
+							imgBufferWidth
```
First of all, width IN BYTES.

Nooope, you don't want a buffer overrun (too small line data width in source), you want to copy the minimum of `textureGliStride` and `region->imageExtent.width`, then divide by format block width rounding up (`(min(.....)+blockWidth-1u)/blockWidth`), followed by multiply by texel/block byte size (we have a function that gives you that given the `E_FORMAT`)

AnastaZIuk replied:
right now I've copied entire row into gli texture without padding in the buffer, but I don't use what you have written (and not sure If my equation is correct therefore)

devshgraphicsprogramming commented on src/irr/asset/CGLIWriter.cpp:
```diff
@@ -153,12 +153,28 @@ namespace irr
 			{
 				const auto ptrBeginningOfRegion = data + region->bufferOffset;
 				const auto layerSize = getFullSizeOfLayer(region->imageSubresource.mipLevel);
```
wrong thing to use, you should be computing that from `region` in its entirety (cause size of layer depends on `bufferRowLength` and `bufferImageHeight` and pixel byte size and eventual depth) also you can compute that without the mip-level at all.

P.S. Check Vulkan Specificationm for semantics of `bufferRowLength` and `bufferImageHeight` when block compressed textures are involved (does the length and height still mean pixels or blocks?)

AnastaZIuk replied:
it's computed like bellow

> wrong thing to use, you should be computing that from region in its entirety (cause size of layer depends on bufferRowLength and bufferImageHeight and pixel byte size and eventual depth) also you can compute that without the mip-level at all.

AnastaZIuk replied:
I will check it out

AnastaZIuk committed:
Avoid cache thrashing, add some formats to the loader

devshgraphicsprogramming commented on src/irr/asset/CGLILoader.cpp:
```diff
@@ -415,57 +417,57 @@ namespace irr
-				case gl::INTERNAL_RGB_ETC: return getTranslatedFinalFormat();						//GL_COMPRESSED_RGB8_ETC1
+				case gl::INTERNAL_RGB_ETC: return getTranslatedFinalFormat(EF_ETC2_R8G8B8_UNORM_BLOCK);						//GL_COMPRESSED_RGB8_ETC1
 				case gl::INTERNAL_RGB_ETC2: return getTranslatedFinalFormat(EF_ETC2_R8G8B8_UNORM_BLOCK);						//GL_COMPRESSED_RGB8_ETC2
 				case gl::INTERNAL_RGBA_PUNCHTHROUGH_ETC2: return getTranslatedFinalFormat();		//GL_COMPRESSED_RGB8_PUNCHTHROUGH_ALPHA1_ETC2
```
still needs doing

AnastaZIuk replied:
done

devshgraphicsprogramming commented on include/irr/asset/IImageLoader.h:
```diff
@@ -33,29 +33,26 @@ namespace irr
+					
+					constexpr auto pixelByteSize = 3;
+					const void* planarData[] = { redChannelDataLayer, nullptr, nullptr, nullptr };
+					uint8_t* out = _IRR_NEW_ARRAY(uint8_t, imageSizeWithPitch.X * imageSizeWithPitch.Y * imageSizeWithPitch.Z);
```
don't allocate memory in the function! just take an `ICPUBuffer` to go with the `region` and use that.

AnastaZIuk committed:
Add a few formats to the Writer

AnastaZIuk committed:
Merge branch 'shaders' of https://github.com/Crisspl/IrrlichtBAW into GLIdep

For fake PR

AnastaZIuk committed:
Prepare to change texture iteration assuming blocks

AnastaZIuk committed:
Take pixelOrBlockSize into account

AnastaZIuk committed:
Perform tests on textures using faces, layers and mipmaps

Everything works for non compressed formats. Now I need to port my loader/writer
to use BC formats as well, because now it crashes (I've used DXT5). It seems to me
I need to do something else with size of texture returned from GLI library and also
change a little bit an equation for copying rows to use blocks.

devshgraphicsprogramming commented on src/irr/asset/CGLIWriter.cpp:
```diff
@@ -166,16 +165,26 @@ namespace irr
+
 				const auto imgBufferWidth = region->bufferRowLength > 0 ? region->bufferRowLength : region->imageExtent.width;
-				const auto imgBufferWidthInBytes = imgBufferWidth * sizeof(singleChannelByteSize);
+				const auto imgBufferWidthInBytes = imgBufferWidth * singleChannelByteSize;
```
stride includes all channels and pixels in a block

devshgraphicsprogramming commented on src/irr/asset/CGLIWriter.cpp:
```diff
@@ -166,16 +165,26 @@ namespace irr
+				const auto imgBufferWidthInBytes = imgBufferWidth * singleChannelByteSize;
 				const auto imgBufferHeight = region->bufferImageHeight > 0 ? region->bufferImageHeight : region->imageExtent.height;
+
+				const auto pixelOrBlockCount = textureGliImgHeight * textureGliStride; // it's wrong, I have to make sure I get correct value
```
that would actually be:
`(imageStrideInTexels/blockSize.x)*(imageHeightInTexels/blockSize.y)*(imageDepth/blockSize.z)`
it would also include the fractional and padding blocks.

GLI and your input region might have different strides and heights, so you'd be getting different block counts.

devshgraphicsprogramming commented on src/irr/asset/CGLIWriter.cpp:
```diff
@@ -186,13 +195,18 @@ namespace irr
 					const auto layerData = texture.data(gliLayer, gliFace, region->imageSubresource.mipLevel);
 					const auto sourceData = ptrBeginningOfRegion + (layer * layerSize);
```
yeah read https://github.com/Crisspl/IrrlichtBAW/pull/21/files/80e64777de0708872d875906cd1439c291307e99..dbf4f204b87c39a090827d50ca0844bd7d3fcb89#r368247906

You need to account for x,y,z offsets in the minimum of a region.

devshgraphicsprogramming commented on src/irr/asset/CGLIWriter.cpp:
```diff
@@ -186,13 +195,18 @@ namespace irr
 					const auto sourceData = ptrBeginningOfRegion + (layer * layerSize);
 
-					for (uint32_t yPos = 0; yPos < textureGliImgHeight; ++yPos)
+					for (size_t index = 0; index < pixelOrBlockCount; ++index)
```
you need 2 separate loops, one for Z and one for Y,

right now you're copying a whoe image row.line for every block/pixel causes huge overwrites.

devshgraphicsprogramming commented on src/irr/asset/CGLIWriter.cpp:
```diff
@@ -186,13 +195,18 @@ namespace irr
-							imgBufferWidth																// get row stride with no pitch
+							reinterpret_cast<uint8_t*>(layerData) + (posYOfTexelOrBlock * textureGliStrideInBytes),	// copy to a beginning of a certain row
+							sourceData + (posYOfTexelOrBlock * imgBufferWidthInBytes),								// get adjusted beginning of a certain row using to copying process
+							textureGliStrideInBytes																    // get gli row stride with no pitch
```
this need to be a minimum of the range between region offset and max coord (extent) and GLIByteStride - region offset 

Obviously with all values converted to bytes, and 3 checks whether all parameters that dea with dimensions and extents are aligned to block size before you even enter the loop.

devshgraphicsprogramming commented on src/irr/asset/CGLIWriter.cpp:
```diff
@@ -166,16 +165,26 @@ namespace irr
 
+			const auto ROW_TEXEL_OR_BLOCK = 0; // TODO
+			const auto blockDimension = getBlockDimensions(imageInfo.format);
+			const auto sourceTexelOrBlockStride = getTexelOrBlockBytesize(imageInfo.format) * ROW_TEXEL_OR_BLOCK;
```
stride is in physical units (multiples of blocks/texels, although to be fair texels are just 1x1x1 blocks) not bytes

devshgraphicsprogramming commented on src/irr/asset/CGLIWriter.cpp:
```diff
@@ -195,18 +189,19 @@ namespace irr
-						);
-					}
+					for (size_t zPos = 0; zPos < textureGLIDepth; ++zPos)
+						for (size_t yPos = 0; yPos < imgBufferHeight; ++yPos)
```
region is source, GLI is sink, iterate over region's defined rectangle

devshgraphicsprogramming commented on src/irr/asset/CGLIWriter.cpp:
```diff
@@ -195,18 +189,19 @@ namespace irr
+							(
+								reinterpret_cast<uint8_t*>(layerData) + shiftAdressingRowGLI,	// copy to a beginning of a certain row
+								sourceData + shiftAdressingRowIRR,								// get adjusted beginning of a certain row using to copying process
+								textureGliStrideInBytes									        // get gli row stride with no pitch
```
still need to work it out from region's extent (and clamp it appropriately so you don't go out of bounds if sink is too small)

devshgraphicsprogramming commented on src/irr/asset/CGLIWriter.cpp:
```diff
@@ -195,18 +189,19 @@ namespace irr
+						for (size_t yPos = 0; yPos < imgBufferHeight; ++yPos)
+						{
+							const size_t shiftAdressingRowGLI = ((zPos * textureGliImgHeight + yPos) * textureGliStride) * texelBlockByteSize;
+							const size_t shiftAdressingRowIRR = ((zPos * imgBufferHeight + yPos) * imgBufferWidth) * texelBlockByteSize;
```
forgetting the offsets in X

AnastaZIuk committed:
Test it and make it work with BC formats

AnastaZIuk committed:
Nearly done, need setting destination offset of data

AnastaZIuk committed:
Revert CGLILoader

devshgraphicsprogramming committed:
test this, it should work

AnastaZIuk committed:
Fix some leaks, test everything one more time

AnastaZIuk committed:
Make use of getMipSize()

AnastaZIuk committed:
Use subresource miplevels

devshgraphicsprogramming commented:
close this PR, merge my shaders branch into this instead
